registry/consul/listener.go

```diff
@@ -0,0 +1,206 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package consul
+
+import (
+	"sync"
+)
+
+import (
+	consul "github.com/hashicorp/consul/api"
+	"github.com/hashicorp/consul/api/watch"
+	perrors "github.com/pkg/errors"
+)
+
+import (
+	"github.com/apache/dubbo-go/common"
+	"github.com/apache/dubbo-go/common/logger"
+	"github.com/apache/dubbo-go/registry"
+	"github.com/apache/dubbo-go/remoting"
+)
+
+// Consul listener wraps the consul watcher, to
+// listen the service information change in consul
+// registry.
+type consulListener struct {
+	// Registry url.
+	registryUrl common.URL
+
+	// Consumer url.
+	consumerUrl common.URL
+
+	// Consul watcher.
+	plan *watch.Plan
+
+	// Most recent service urls return by
+	// watcher.
+	urls []common.URL
+
+	// All service information changes will
+	// be wrapped into ServiceEvent, and be
+	// sent into eventCh. Then listener's
+	// Next method will get event from eventCh,
+	// and return to upstream.
+	eventCh chan *registry.ServiceEvent
+
+	// All errors, happening in the listening
+	// period, will be caught and send into
+	// errCh. Then listener's Next method will
+	// get error from errCh, and return to
+	// upstream.
+	errCh chan error
+
+	// Done field represents whether consul
+	// listener has been closed. When closing
+	// listener, this field will be closed,
+	// and will notify consul watcher to close.
+	done chan struct{}
+
+	// After listener notifies consul watcher
+	// to close, listener will call wg.wait to
+	// make sure that consul watcher is closed
+	// before the listener closes.
+	wg sync.WaitGroup
+}
+
+func newConsulListener(registryUrl common.URL, consumerUrl common.URL) (*consulListener, error) {
+	params := make(map[string]interface{}, 8)
+	params["type"] = "service"
+	params["service"] = consumerUrl.Service()
+	params["tag"] = "dubbo"
+	params["passingonly"] = true
+	plan, err := watch.Parse(params)
+	if err != nil {
+		return nil, err
+	}
+
+	listener := &consulListener{
+		registryUrl: registryUrl,
+		consumerUrl: consumerUrl,
+		plan:        plan,
+		urls:        make([]common.URL, 0, 8),
+		eventCh:     make(chan *registry.ServiceEvent),
```

Comment:
zero length chan?

Reply:
@AlexStocks Em, I am not sure what length that I should initialize with? Is there any number for advice.

Reply:
maybe 32 is enough.

Reply:
@AlexStocks I follow your advice and add it in new commit.

Reply:
ok